### ISSUE TITLE
Building with 32bit OCaml: Integer literal exceeds the range of representable integers of type int

### ISSUE THREAD

**Issue by adamss3** [FStarLang/FStar#1332]
I'm hoping that there is a DEB package for an i386 cpu.  I tried to build from source, but there are issues getting o'caml packages installed.

**Comment by karthikbhargavan** [comment]
I am curious, what kind of machine are you trying to install it on?
If it is running 32-bit Ubuntu because it is an older CPU, then I would warn you against installing F*,
because the typechecker, even in lax-mode, is quite resource heavy and needs a powerful machine.

**Comment by adamss3** [comment]
It's a Compaq V2000 with

model name    : Mobile AMD Sempron(tm) Processor 2800+
stepping    : 2
cpu MHz        : 1600.000

**Comment by catalin-hritcu** [comment]
@adamss3 F* should properly install on 32 bit Linux systems. Can you explain what's the problem you got installing OCaml packages?

**Comment by adamss3** [comment]
Here is the first error:

#=== ERROR while installing camlp4.4.02+system 
================================#
"install": command not found.

This is from:

opam install ocamlbuild ocamlfind batteries stdint zarith yojson 
fileutils pprint menhir ulex

**Comment by catalin-hritcu** [comment]
@adamss3 What ocaml version are you on? Is it installed via opam? What does `opam switch` say?

**Comment by adamss3** [comment]
Sorry, I was away for a couple of days.  ocaml was installed via 
synaptic package manager.

adamss@v2000e:~$ ocaml -version
The OCaml toplevel, version 4.02.3
adamss@v2000e:~$ opam switch
--     -- 3.07    Official 3.07 release
--     -- 3.08.0  Official 3.08.0 release
--     -- 3.08.1  Official 3.08.1 release
--     -- 3.08.2  Official 3.08.2 release
--     -- 3.08.3  Official 3.08.3 release
--     -- 3.08.4  Official 3.08.4 release
--     -- 3.09.0  Official 3.09.0 release
--     -- 3.09.1  Official 3.09.1 release
--     -- 3.09.2  Official 3.09.2 release
--     -- 3.09.3  Official 3.09.3 release
--     -- 3.10.0  Official 3.10.0 release
--     -- 3.10.1  Official 3.10.1 release
--     -- 3.10.2  Official 3.10.2 release
--     -- 3.11.0  Official 3.11.0 release
--     -- 3.11.1  Official 3.11.1 release
--     -- 3.11.2  Official 3.11.2 release
--     -- 3.12.0  Official 3.12.0 release
--     -- 3.12.1  Official 3.12.1 release
--     -- 4.00.0  Official 4.00.0 release
--     -- 4.00.1  Official 4.00.1 release
--     -- 4.01.0  Official 4.01.0 release
--     -- 4.02.0  Official 4.02.0 release
--     -- 4.02.1  Official 4.02.1 release
--     -- 4.02.2  Official 4.02.2 release
--     -- 4.02.3  Official 4.02.3 release
--     -- 4.03.0  Official 4.03.0 release
--     -- 4.04.0  Official 4.04.0 release
--     -- 4.04.1  Official 4.04.1 release
--     -- 4.04.2  Official 4.04.2 release
--     -- 4.05.0  Official 4.05.0 release
system  C system  System compiler (4.02.3)
# 152 more patched or experimental compilers, use '--all' to show

**Comment by catalin-hritcu** [comment]
@adamss3 Can you please try this again after running?
```
$ opam install 4.02.3
$ eval `opam config env`
```

**Comment by adamss3** [comment]
After those two steps:


adamss@v2000e:~/FStar-master$ make
make -C src/ocaml-output
make[1]: Entering directory '/home/adamss/FStar-master/src/ocaml-output'
cp "../extraction/ml/FStar_Extraction_ML_PrintML_4.02.X.ml" 
../extraction/ml/FStar_Extraction_ML_PrintML.ml
fatal: Not a git repository (or any of the parent directories): .git
Not a git repository
To compare two paths outside a working tree:
usage: git diff [--no-index] <path> <path>
echo 'open FStar_Util' > FStar_Version.ml
echo 'let dummy () = ();;' >> FStar_Version.ml
echo 'FStar_Options._version := "0.9.5.0";;' >> FStar_Version.ml
echo 'FStar_Options._platform := "Linux_i686";;' >> FStar_Version.ml
echo 'FStar_Options._compiler := "OCaml 4.02.3";;' >> FStar_Version.ml
echo 'FStar_Options._date := "2017-11-09T21:01:52-05:00";;' >> 
FStar_Version.ml
echo 'FStar_Options._commit:= " (dirty)";;' >> FStar_Version.ml
cd ../../ && ocamlbuild -cflag -g -I src/ocaml-output -I src/basic/ml -I 
src/parser/ml -I src/fstar/ml -I src/extraction/ml -I src/prettyprint/ml 
-I src/tactics/ml -I ulib/ml -j 24 -build-dir src/ocaml-output/_build 
-use-ocamlfind main.native
+ ocamlfind ocamldep -package stdint -package compiler-libs -package 
compiler-libs.common -package menhirLib -package dynlink -package pprint 
-package ulex -package yojson -package batteries -package zarith 
-modules src/fstar/ml/main.ml > src/fstar/ml/main.ml.depends
ocamlfind: Package `stdint' not found
Command exited with code 2.
Compilation unsuccessful after building 1 target (0 cached) in 00:00:00.
Makefile:104: recipe for target 'fstar.exe' failed
make[1]: *** [fstar.exe] Error 10
make[1]: Leaving directory '/home/adamss/FStar-master/src/ocaml-output'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by catalin-hritcu** [comment]
@adamss3 You still need your original command before trying to build F*
> $ opam install ocamlbuild ocamlfind batteries stdint zarith yojson fileutils pprint menhir ulex

The question is whether it works better with the OCaml version from OPAM.

**Comment by adamss3** [comment]
I ran that command and the output changed, but there is still a show 
stopping error:

make -C src/ocaml-output
make[1]: Entering directory '/home/adamss/FStar-master/src/ocaml-output'
cp "../extraction/ml/FStar_Extraction_ML_PrintML_4.02.X.ml" 
../extraction/ml/FStar_Extraction_ML_PrintML.ml
fatal: Not a git repository (or any of the parent directories): .git
Not a git repository
To compare two paths outside a working tree:
usage: git diff [--no-index] <path> <path>
echo 'open FStar_Util' > FStar_Version.ml
echo 'let dummy () = ();;' >> FStar_Version.ml
echo 'FStar_Options._version := "0.9.5.0";;' >> FStar_Version.ml
echo 'FStar_Options._platform := "Linux_i686";;' >> FStar_Version.ml
echo 'FStar_Options._compiler := "OCaml 4.02.3";;' >> FStar_Version.ml
echo 'FStar_Options._date := "2017-11-10T10:31:27-05:00";;' >> 
FStar_Version.ml
echo 'FStar_Options._commit:= " (dirty)";;' >> FStar_Version.ml
cd ../../ && ocamlbuild -cflag -g -I src/ocaml-output -I src/basic/ml -I 
src/parser/ml -I src/fstar/ml -I src/extraction/ml -I src/prettyprint/ml 
-I src/tactics/ml -I ulib/ml -j 24 -build-dir src/ocaml-output/_build 
-use-ocamlfind main.native
+ ocamlfind ocamldep -package stdint -package batteries -package zarith 
-modules ulib/ml/FStar_Int32.ml > ulib/ml/FStar_Int32.ml.depends
File "ulib/ml/FStar_Int32.ml", line 13, characters 33-43:
Error: Integer literal exceeds the range of representable integers of 
type int
Command exited with code 2.
Compilation unsuccessful after building 17 targets (0 cached) in 00:00:01.
Makefile:104: recipe for target 'fstar.exe' failed
make[1]: *** [fstar.exe] Error 10
make[1]: Leaving directory '/home/adamss/FStar-master/src/ocaml-output'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by catalin-hritcu** [comment]
Now we're talking, seems that `ulib/ml/FStar_Int32.ml` is broken on 32 bit machines. The code looks like this:
```
let add_mod a b = (add a b) land 4294967295
```
@ad-l @jkzinzindohoue @victor-dumitrescu Do you have any ideas?

**Comment by adamss3** [comment]
I'm certainly no ocaml expert but what I see is that the integer value 
is a max 32 bit unsigned value.  What will be the default type for add mod?

**Comment by msprotz** [comment]
You could try overriding the files in `ulib/ml` with those from `ulib/ml/32bit`, but that hasn't been tested in a while and is probably broken.

I really don't think we have good support for 32-bit builds; last time I did it, it took me several hours of manual work to make mitls.exe happen in 32-bit mode.

**Comment by adamss3** [comment]
I copied up the files from 32bit, and it built with no errors. I'll be 
testing it now.  Thanks.

**Comment by mtzguido** [comment]
This should be fixed when we merge #1850

**Comment by mtzguido** [comment]
Now merged.
